### ISSUE TITLE
Ability to Renew/Refresh/Extend Locks

### ISSUE THREAD

**Issue by candrews** [lukas-krecan/ShedLock#151]
**Is your feature request related to a problem? Please describe.**
I have a use case where I'd like to keep the lock even after it expires if possible. In other words, I'd like to renew/refresh/extend a lock that is currently being held.

**Describe the solution you'd like**
A few potential approaches includes
* Add a `boolean extend(Instant lockAtMostUntil, Instant lockAtLeastUntil)` method to `net.javacrumbs.shedlock.core.SimpleLock`
* Add a `Optional<SimpleLock> net.javacrumbs.shedlock.core.LockProvider.extend(SimpleLock lock, Instant lockAtMostUntil, Instant lockAtLeastUntil)` to `net.javacrumbs.shedlock.core.LockProvider`

Such methods would have to be called with locks that are currently being held (not expired or unlocked).

**Describe alternatives you've considered**
Currently, I have to do something like this:
```
@Scheduled(fixedRate=60000)
public void lock() throws Exception {
	boolean everAcquiredLock = false;
	while(! shuttingDown) {
		final LockConfiguration lockConfiguration = new LockConfiguration(
				"lock",
				Instant.now().plus(Duration.ofMillis(60000))); 
		final Optional<SimpleLock> optionalLock = lockProvider.lock(lockConfiguration);
		if(optionalLock.isPresent()) {
			try {
				if(! everAcquiredLock) {
					everAcquiredLock = true;
					log.info("Acquired lock, processing");
					startLongRunningTaskInOtherThread(); // TODO do the lock protected background work here
				}
				Thread.sleep(60000);
			}finally {
				log.info("Releasing lock");
				optionalLock.get().unlock();
			}
		}else {
			break;
		}
	}
	if(everAcquiredLock) {
		log.info("Lock not re-acquired, stopping processing");
		startLongRunningTaskInOtherThread(); // TODO stop the background work here
	}
}
```

That's not great though, as:
* It's complicated
* There's a window where another system could steal the lock resulting in unnecessary lock passing
* There's a window where the current lock holder could be waiting for its long running task to end after it's given up the lock and another system acquires the lock and start its long running task, resulting in long running tasks running on two systems (preventing that is the whole point of the lock)

**Additional context**

**Comment by lukas-krecan** [comment]
Hi, thanks for the feedback and sorry for the delay. 

What is the use-case for the extension? I am quite hesitant to add it since it will rise the complexity of all the lock providers. 

Maybe we can add a new interface extending the SimpleLock and implement it only in some LockProviders, which LockProvider are do you use?

**Comment by candrews** [comment]
I think I gave the use case in the description of this issue - I have a long running job that can be run in parts (it is interruptable) but is more efficient if run uninterrupted so I'd like to only run once globally.

I'm using `JdbcLockProvider`.

However, I'd love to see it for all lock providers... one of the nice things about shedlock is that the lock implementation can be easily and transparently switched.

**Comment by lukas-krecan** [comment]
What about something like this https://github.com/lukas-krecan/ShedLock/pull/160/files#diff-3367643514070c7e64f9bb823f19021eR24?

**Comment by candrews** [comment]
> What about something like this https://github.com/lukas-krecan/ShedLock/pull/160/files#diff-3367643514070c7e64f9bb823f19021eR24?

```
    default Optional<SimpleLock> extend(Instant lockAtMostUntil, Instant lockAtLeastUntil) {
        throw new UnsupportedOperationException();
    }
```
From a user perspective, I don't think there's any difference between a lock provider than can't extend and a failed attempt to extend, so I don't think there's a need to throw `UnsupportedOperationException`. Instead of throwing an exception, perhaps the default could be to simply not extend the lock? So it becomes:
```
    default Optional<SimpleLock> extend(Instant lockAtMostUntil, Instant lockAtLeastUntil) {
        return Optional.empty();
    }
```

**Comment by lukas-krecan** [comment]
I see your point. In my opinion there should be a difference between a lock that can not be extended because it's already expired and a lock that can not be extended because it's not supported.

If you think it's useful as it is, I will release it so it can be tested in real life.

**Comment by candrews** [comment]
> If you think it's useful as it is, I will release it so it can be tested in real life.

That's a big :+1: from me - thank you!

**Comment by lukas-krecan** [comment]
Released in 2.6.0
